### PR TITLE
fix(smtp): guard null-MX domain in Send to avoid nil panic (#379)

### LINKED ISSUE #379: smtp: null-MX nil panic, 4xx short-circuit, missing MX cache, no ctx
## Problem

Four issues in \`internal/smtp/sender.go\` and the surrounding SMTP path:

### 6.1 — Null-MX nil-deref panic (CRITICAL)

\`internal/smtp/sender.go:80-95\`:

\`\`\`go
var lastErr *smtpError
for _, mx := range mxs {
    err := deliver(from, to, msg, mx, false, s.Hostname)
    if err == nil { return nil }
    if err.code > 200 {
        return err
    }
    lastErr = err
}
err = fmt.Errorf(\"all MXs failed, last error: %w\", lastErr)
return newSMTPError(err, false, lastErr.Code())   // <-- panic if mxs was empty
\`\`\`

\`lookupMXs\` can legitimately return \`(nil, nil)\` for domains with explicit RFC 7505 \"null MX\" (a \`0 .\` record). The for-loop is skipped, \`lastErr\` stays \`nil\`, then \`lastErr.Code()\` panics → crashes the SMTPSender goroutine and (depending on Parallel pool behaviour) potentially the worker.

### 6.2 — MX 4xx short-circuit prevents fallback

Same loop: \`if err.code > 200 { return err }\`. Any MX returning a 4xx (450, 451, 452 — *transient* per RFC 5321) immediately aborts the loop and returns that error as if it were terminal for all MXs. Multi-MX recipients lose retry diversity — they should fall through to the next MX.

### 6.3 — No MX cache → DNS lookup per send

\`internal/smtp/sender.go:75\` → \`lookupMXs(toDomain)\` calls \`net.LookupMX\` per delivery. For batches targeting the same recipient domain (very common: 10k recipients on \`gmail.com\`), this is 10k redundant DNS queries.

### 6.4 — \`Send\` has no ctx; shutdown blocks 120 s × in-flight

\`internal/smtp/smtp.go:5-7\` and \`internal/smtp/sender.go:67\` — \`Send(from, to, msg)\` has no \`context.Context\`. Worker shutdown (\`pkg/smtpsender/smtpsender.go:90\`) calls \`tasks.WaitAndClose()\` which blocks up to \`smtpTotalTimeout=120s\` per in-flight job. \`net.DialTimeout\` is used but cannot be cancelled.

## Severity

- 6.1: **CRITICAL** (panic)
- 6.2: **MED** (delivery diversity loss)
- 6.3: **MED** (cost / latency)
- 6.4: **MED** (operational — slow rolling deploys)

### 6.1
Handle empty \`mxs\` early, before the loop:

\`\`\`go
mxs, lerr := lookupMXs(toDomain)
if lerr != nil { return lerr }
if len(mxs) == 0 {
    return newSMTPError(errors.New(\"recipient domain has null MX\"), true, 550)
}
\`\`\`

### 6.2
Only short-circuit on 5xx (permanent) responses; let 4xx fall through to next MX:

\`\`\`go
if err.code >= 500 {
    return err
}
lastErr = err
\`\`\`

### 6.3
Add a small TTL cache (5 min) keyed by recipient domain → []mx, with \`singleflight\` to dedupe concurrent lookups:

\`\`\`go
type mxCache struct {
    sf singleflight.Group
    mu sync.RWMutex
    entries map[string]mxEntry
}
\`\`\`

Lives in \`internal/smtp/\` and is constructed once per \`sender\`.

### 6.4
Plumb \`ctx context.Context\` through \`smtp.Sender.Send\` and \`deliver\`. Use \`(&net.Dialer{}).DialContext(ctx, ...)\` and \`conn.SetDeadline\` derived from \`ctx.Deadline()\`. The \`Parallel\` pool callback already has access to a ctx — pass it down.

## Acceptance criteria

- [ ] Empty MX list returns a permanent error instead of panicking
- [ ] 4xx from one MX falls through to next MX in the list
- [ ] MX cache hit ratio > 90% under steady-state batch load (test by counting \`net.LookupMX\` calls in a fixture)
- [ ] \`Sender.Send\` accepts \`context.Context\`; cancellation interrupts dial / connection
- [ ] Existing \`internal/smtp\` tests still pass; add a test for the null-MX case

## Effort / impact

M effort total (~3 days), HIGH impact when bundled. 6.1 alone is S effort and should ship immediately even ahead of the bundle.

## Notes for the implementer

- 6.1 is a 4-line fix and could land as a hotfix PR before the rest
- The MX cache should be opt-out via config in case operators want strict per-send DNS for forensic reasons
- For 6.4, mind that \`net/smtp\` exposes \`smtp.NewClient(conn, ...)\` not a \`Dialer\` — you'll need to dial first then \`smtp.NewClient\`, which is already the current pattern in \`deliver\`

### PR DESCRIPTION
**Pull request by SAY-5**
**What this PR does / why we need it**:
Addresses 6.1 in #379. `lookupMXs` can legitimately return `([], nil)` for domains that publish an RFC 7505 null MX record (`0 .`). The for-loop in `Send` was then skipped, leaving `lastErr == nil`, and the trailing `lastErr.Code()` panicked, taking out the SMTPSender goroutine.

Extracts the MX-iteration loop into `sendToMXs(...)` and returns a permanent 550 when `mxs` is empty, mirroring the recommendation in the issue. Adds `internal/smtp/sender_test.go` with two cases (`nil` slice and `[]string{}`) that both panic on `main` and pass after the fix.

Scoped to 6.1 only, 6.2/6.3/6.4 left for follow-ups per the issue's "could land as a hotfix PR before the rest" note.

**Which issue(s) this PR fixes**
Fixes #379 (item 6.1).

### CODE REVIEW

File: internal/smtp/sender.go [hunk 1]
```diff
@@ -77,9 +77,17 @@ func (s *sender) Send(from, to string, msg []byte) SenderError {
 		return lerr
 	}
 
+	return sendToMXs(from, to, msg, mxs, s.Hostname)
+}
+
+func sendToMXs(from, to string, msg []byte, mxs []string, hostname string) *smtpError {
```
Comment (ludusrusso) [inline internal/smtp/sender.go]: this introduce a bug, sendToMXs returns the real imlpementation, while Send cast it to the interface SenderError. 

Due to the cast, SenderError is always not nil and may introduce bugs. Check here

Comment (ludusrusso) [reply internal/smtp/sender.go]:
```go
func TestSendToMXs_TypedNilInterfaceBug(t *testing.T) {
	var iface SenderError = sendToMXsNilSuccess()
	if iface != nil {
		t.Fatalf("expected nil interface on success, got non-nil (type=%T value=%v)", iface, iface)
	}
}

// Mirrors the success return path inside sendToMXs.
func sendToMXsNilSuccess() *smtpError {
	return nil
}
```

**Review by ludusrusso — CHANGES_REQUESTED** [review]

**Review by ludusrusso — APPROVED** [review]
